### DISCUSSION TITLE
file magic for sysstat

### DISCUSSION THREAD

**Discussion by rowlap** [sysstat/sysstat#297] [Ideas]
As `file(1)` currently prints "data" for the daily data file, looking at sa.h we find SYSSTAT_MAGIC and FORMAT_MAGIC which would be a good starting point for teaching `file` to recognise binary format.

I'm happy to spend time creating the "magic" rules to do this, but thought it worth asking first whether it's been done already, or is hard for reasons I haven't spotted, or ...?

**Comment by sysstat** [comment]
I agree: This is a good idea and it has never been done before :)
Please note that FORMAT_MAGIC value will change when the internal format of the daily data file is no longer compatible with that of previous versions. So `file(1)` should use only **SYSSTAT_MAGIC** to recognise a sysstat data file.

**Comment by rowlap** [comment]
I gave this a first pass:
````
$ cat sysstat.magic
# SYSSTAT_MAGIC / struct file_magic introduced in sysstat version 8.1.1
# these rules will not work on older files

0 short 0xd596 sysstat data file,
# hex format string mirrors sa_open_read_magic debug output
>2 short x format %x, created by
>4 byte x version %u.
>5 byte x \b%u.
>6 byte x \b%u
# omit file_magic.sysstat_extraversion?

$ file -m sysstat.magic /var/log/sa/sa19
/var/log/sa/sa19: sysstat data file, format 2171, created by version 10.1.5
````

- Seems OK to omit files pre-8.1.1 (2008-02-10).
- Probably fine if files are generated and analysed on the same host, but I'm not sure what this will do with non-native endian (swapped) data
- Do you have a bunch of old sysstat files lying around to try this on?
- I did use FORMAT_MAGIC because it's in the sprit / level of detail that "file" usually produces

**Comment by sysstat** [comment]
I usually set to 9.1.6 the number of the oldest supported sysstat version. So you can probably even omit files pre-9.1.6 (2010-11-10).
Here is a bunch of sysstat files to try:
[data.zip](https://github.com/sysstat/sysstat/files/7571601/data.zip)
One of them (`data-ppc-11.5.5`) has been created on a big endian system.
Thank you for your work.

**Comment by rowlap** [comment]
It's not really a question of "supporting" 8.1.1 / 9.1.6 / any specific version, rather than when did `struct file_magic` appear? That's what the magic file recognises.

For the endian support, why don't we start with native and if there's demand for the byte-swapped rules, they can be added later.

The demo files all worked fine, thanks, except the big endian (as expected).

@sysstat are you happy if I submit these rules to the `file` maintainer? I won't present them as endorsed or "offical", but will just give a link instead to this discussion.

**Comment by sysstat** [comment]
I agree with you, don't bother with big endian support for now.
You can send the rules to the `file` maintainer.
Thx again

**Comment by rowlap** [comment]
I never followed up here for 3 years, but in summary the prefix [isn't considered long enough](https://bugs.astron.com/view.php?id=301) by the `file` maintainer. We could change `SYSSTAT_MAGIC` to be longer, but this doesn't feel like a good enough reason to do so, potentially breaking format compatibility with older versions.

Another approach would be to take `SYSSTAT_MAGIC` and `FORMAT_MAGIC` together, to make 4 bytes. As there are only currently 2 known values for FORMAT_MAGIC (2171 and 2173), this wouldn't be a long list.

**Comment by sysstat** [comment]
Yes,taking both `SYSSTAT_MAGIC` and `FORMAT_MAGIC` could be a solution.
There are 3 possible values for FORMAT_MAGIC: , **0x2171** (starting with sysstat version 9.1.6 dated 2010-11-10), **0x2173** (starting with sysstat version 10.3.1 dated 2014-03-18) and **0x2175** (starting with version 11.7.1 dated 2018-01-12).

**Comment by natoscott** [comment]
Oh, I didn't notice this conversation previously.  This should be working now though:
https://github.com/file/file/commit/61d32c41719a1727a328fd88d4db419f67d95871
